common/scala/src/main/scala/org/apache/openwhisk/core/database/mongodb/MongoDBAsyncStreamSink.scala

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.openwhisk.core.database.mongodb
+
+import java.nio.ByteBuffer
+
+import akka.Done
+import akka.stream.{Attributes, IOResult, Inlet, SinkShape}
+import akka.stream.scaladsl.Sink
+import akka.stream.stage.{AsyncCallback, GraphStageLogic, GraphStageWithMaterializedValue, InHandler}
+import akka.util.ByteString
+import org.mongodb.scala.Completed
+import org.mongodb.scala.gridfs.{AsyncOutputStream}
+
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.util.{Failure, Success, Try}
+
+class MongoDBAsyncStreamSink(stream: AsyncOutputStream)(implicit ec: ExecutionContext)
+    extends GraphStageWithMaterializedValue[SinkShape[ByteString], Future[IOResult]] {
+  val in: Inlet[ByteString] = Inlet("AsyncStream.in")
+
+  override val shape: SinkShape[ByteString] = SinkShape(in)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[IOResult]) = {
+    val ioResultPromise = Promise[IOResult]()
+    val logic = new GraphStageLogic(shape) with InHandler {
+      handler =>
+      var buffers: Iterator[ByteBuffer] = Iterator()
+      var writeCallback: AsyncCallback[Try[Int]] = _
+      var closeCallback: AsyncCallback[Try[Completed]] = _
+      var position: Int = _
+      var writeDone = Promise[Completed]
+
+      setHandler(in, this)
+
+      override def preStart(): Unit = {
+        //close operation is async and thus requires the stage to remain open
+        //even after all data is read
+        setKeepGoing(true)
+        writeCallback = getAsyncCallback[Try[Int]](handleWriteResult)
+        closeCallback = getAsyncCallback[Try[Completed]](handleClose)
+        pull(in)
+      }
+
+      override def onPush(): Unit = {
+        buffers = grab(in).asByteBuffers.iterator
+        writeDone = Promise[Completed]
+        writeNextBufferOrPull()
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        //Work done perform close
+        //Using async "blessed" callback does not work at this stage so
+        // need to invoke as normal callback
+        //TODO Revisit this
```

Comment:
Do we need an open issue here to track? I can see this TODO get lost

Reply:
This is not related to current master code, so I think it is unfit to record as an issue?